### ISSUE TITLE
potential access to freed memory in dbuf_free_range

### ISSUE THREAD

**Issue by avg-I** [zfsonlinux/zfs#9141]
Sorry for not following the template.
This _potential_ and _low impact_ issue was found by a code review of the latest ZoL code (after an unrelated accident with a different ZFS code base).

This is the abbreviated code in question:
```.c
        db = avl_nearest(&dn->dn_dbufs, where, AVL_AFTER);

        for (; db != NULL; db = db_next) {
                db_next = AVL_NEXT(&dn->dn_dbufs, db);

                if (db->db_level != 0 || db->db_blkid > end_blkid) {
                        break;
                }
...
                if (zfs_refcount_count(&db->db_holds) == 0) {
                        ASSERT(db->db_buf);
                        dbuf_destroy(db);
                        continue;
                }
...
        }
```

Let's assume the following scenario.
`db` is an L0 dbuf. It has zero holds.
`db_next`, which is the next dbuf on `dn_dbufs`, is an L1 dbuf that is also a _parent_ of `db`.
Additionally, `db_next` has only a single hold, as a parent, from `db` .
Also, `db_next->db_pending_evict` is true.

Then:
- `dbuf_destroy(db)` will call `dbuf_rele_and_unlock(parent, db, B_TRUE)` where `parent` is `db_next`
- `dbuf_rele_and_unlock` will see that the parent has no remaining holds
- `dbuf_rele_and_unlock` will see that `db_pending_evict` is true
- thus, `dbuf_rele_and_unlock` will call `dbuf_destroy` on the parent
- as a result, the next iteration of the loop in `dbuf_free_range` will see a destroyed dbuf

Most of the time that should not be a problem, however.
The dbuf is freed to the kmem cache (`dbuf_kmem_cache`) and most of its fields are preserved (see `dbuf_dest`).
So, the memory remains accessible and the loop will be aborted after `db->db_level != 0` check.

But if the dbuf is immediately reused and its new level is zero and its new offset is within the range being freed, then that unrelated reincarnation can get processed as if it belonged to the original object.

**Comment by ahrens** [comment]
I agree with your analysis.  In addition, this could cause the recursion problem that the `evicting` argument is meant to prevent.  It seems like if `evicting` is set, we should not call dbuf_destroy() on the parent (but the fix may be more involved since we need to make sure it gets called at some point).
```
/*
 * dbuf_rele() for an already-locked dbuf.  This is necessary to allow
 * db_dirtycnt and db_holds to be updated atomically.  The 'evicting'
 * argument should be set if we are already in the dbuf-evicting code
 * path, in which case we don't want to recursively evict.  This allows us to
 * avoid deeply nested stacks that would have a call flow similar to this:
 *
 * dbuf_rele()-->dbuf_rele_and_unlock()-->dbuf_evict_notify()
 *	^						|
 *	|						|
 *	+-----dbuf_destroy()<--dbuf_evict_one()<--------+
 *
 */
```
